src/casp/services/api/services/consensus_engine/strategies.py

```diff
@@ -0,0 +1,230 @@
+        :param knn_query: The result of the kNN query.
+
+        :return: A list of summarized query neighbor context annotations.
+        """
+        ...
+
+
+class ConsensusStrategyType(str, Enum):
+    """
+    Enum representing the available methods for consensus engine calculations. Methods include:
+
+    - ``ONTOLOGY_AWARE``: Utilizes cell type ontology to enrich context summarization.
+    - ``CELL_TYPE_STATISTICS``: Counts cell types and distance statistics within the neighborhood of a query to
+        summarize context.
+    """
+
+    ONTOLOGY_AWARE = "ontology_aware"
+    CELL_TYPE_STATISTICS = "cell_type_statistics"
+
+
+class CellTypeStatisticsConsensusStrategy(ConsensusStrategyProtocol):
+    """
+    Handle cell type count consensus strategy, summarizing query neighbor context by cell type distribution.
+    """
+
+    def __init__(
+        self, cell_operations_dm: CellOperationsDataManager, cas_model: models.CASModel, include_dev_metadata: bool
+    ):
+        self.cell_operations_dm = cell_operations_dm
+        self.cas_model = cas_model
+        self.include_dev_metadata = include_dev_metadata
+
+    def summarize(self, query_ids: t.List[str], knn_query: MatchResult) -> schemas.QueryAnnotationCellTypeCountType:
+        """
+        Summarize query neighbor context by cell type distribution, querying a database for the distribution.
+
+        :param query_ids: List of query cell IDs.
+        :param knn_query: The result of the kNN query.
+
+        :return: An instance of `schemas.QueryAnnotationCellTypeCountType`, representing the summarized cell type
+            distribution of query neighbors.
+        """
+        temp_table_fqn = self.cell_operations_dm.insert_matches_to_temp_table(
+            query_ids=query_ids, knn_response=knn_query
+        )
+
+        if self.include_dev_metadata:
+            return self.cell_operations_dm.get_neighborhood_distance_summary_dev_details(
+                cas_model=self.cas_model, match_temp_table_fqn=temp_table_fqn
+            )
+
+        return self.cell_operations_dm.get_neighborhood_distance_summary(
+            cas_model=self.cas_model, match_temp_table_fqn=temp_table_fqn
+        )
+
+
+class CellOntologyResource:
+    """
+    Handles cell ontology resource data, such as the ancestors dictionary and cell ontology term ID mappings.
+
+    Attributes:
+        ancestors_dictionary: Maps cell ontology term IDs to their ancestor IDs.
+        ontology_term_id_to_name_dict: Maps cell ontology term IDs to cell type names.
+
+    Raises:
+        ValueError: If either ``ancestors_dictionary`` or ``cell_ontology_term_id_to_cell_type`` is missing from the
+            provided dictionary.
+    """
+
+    def __init__(self, cell_ontology_resource_dict: t.Optional[t.Dict[str, t.Any]] = None):
+        if cell_ontology_resource_dict is None:
+            with open(settings.GCS_CELL_ONTOLOGY_RESOURCE_FILE, "rb") as f:
```

Comment:
Cell ontology is actually a rapidly evolving resource:
https://github.com/obophenotype/cell-ontology

Here are the releases:
https://github.com/obophenotype/cell-ontology/releases

CELLxGENE is expected to keep migrating to newer versions, and they are actively involved in extending this ontology. While it is unlikely that they update existing cell type annotations for datasets they have already ingested (though it might happen, and we should ask them how to some updates from them), newly ingested datasets will use newer versions of their schema which would use newer cell ontology releases, which would include both new terms and new relationships. Each specific version of the CELLxGENE schema is pegged to a specific release date of the cell ontology. Cell ontology releases are tagged by their release date.

How can we use combine queries across different referrence datasets that use different cell ontology version? It is likely not a very fundamental problem, since most (if not all) new ontology releases are extensions of previous releases. Probably some terms will be deprecated and eventually removed. A proper solution here is to have a cell ontology migration tool (e.g. to replace deprecated terms with preferred terms), so that we migrate all of the data in our DataStore to the latest cell ontology release. Another solution would be to perform migration on-the-fly here (if we know neighbor X comes from dataset Y with cell ontology Z, we can migrate the ontology metadata of that neighbor to the newest cell ontology release on-the-fly). Let's make an issue for this for further discussions and settle for a simpler solution here (see below).

Let us do the following:

- Include scripts in the repo (or better yet, a protected API) to generate the parsed cell ontology JSON from a given versioned release and store it here:
```
gs://{PROJECT_BUCKET_NAME}/consensus_engine/parsed_cell_ontology_release_vYYYY-MM-DD.json
```
All versioned releases can be downloaded from:
```
https://github.com/obophenotype/cell-ontology/releases/download/vYYYY-MM-DD/cl.owl
```

- We expose which version of the cell ontology is to be used in `CellOntologyResource`. If the (parsed) version exists (`gs://{PROJECT_BUCKET_NAME}/consensus_engine/parsed_cell_ontology_release_vYYYY-MM-DD.json`) we proceed. Otherwise, we return an exception. In the future, we can register parsed cell ontologies as a separate table + API call to let the users know which versions of the ontology we have in store (make an issue).

- When we run the consensus-aware scoring, it is possible that we may not have some of the ontology terms in the loaded resource (e.g. if we are using an older resource and but the neighbors are coming from a newer dataset using a newer resource). For now, we ignore those neighbors and keep track of the number of ignored neighbors (see below).

Reply:
Let's create an issue and work on it through a separate PR. I will include more documentation on cell ontology to keep track of how the resource was created within this PR.

Reply:
It just seems like we need to talk more about the versioning of cell ontology and how to better keep track of the records, how this would affect our data and models as well.